content/en/docs/Creating derivatives/build_iso.md

```diff
@@ -76,182 +47,128 @@ label: "COS_LIVE"
-    type: docker
+repositories:
+  - uri: quay.io/costoolkit/releases-green
 
 ```
 
-Flags:
-- **local**: Path to local luet repository to use to install packages from
-- **image**: Optional image to use as rootfs
-
-## Configuration reference
-
-### `rootfs_image`
-
-A container image reference (e.g. `quay.io/.../...:latest`) that will be used as a rootfs for the ISO.
-
-### `packages.rootfs`
+Packages or sources can be a Luet package (as in the example), an image reference (then an explicit tag is required) or a local path. Sources are stacked in the given order, so one can easily overwrite or append data by simply adding a local path as the last source.
 
-A list of luet packages to install in the rootfs. The rootfs will be squashed to a `rootfs.squashfs` file
+### Command flags
 
-### `packages.uefi`
+- **name**: Name of the ISO image. It will be used to generate the `*.iso` file name
+- **output**: Path of the destination folder of created images
+- **date**: If present it includes the date in the generated file name
+- **overlay-rootfs**: Sets the path of a tree to overlay on top of the system root-tree
+- **overlay-uefi**: Sets the path of a tree to overaly on top of the EFI image root-tree
+- **overlay-iso**: Sets the path of a tree to overlay on top of the ISO filesystem root-tree
+- **label**: Sets the volume label of the ISO filesystem
+- **repo**: Sets the URI of a repository to include together with the repositores set in manifest or the default one if no repositories are set in manifest. This option can be set multiple times.
 
-A list of luet packages to be present in the efi ISO sector.
-
-### `packages.isoimage`
-
-A list of luet packages to be present in the ISO image.
-
-### `repository.packages`
-
-A list of package repository (e.g. `repository/mocaccino-extra`) to be installed before `luet install` commands Requirements
+## Configuration reference
 
-### `initramfs.kernel_file`
+### `iso.rootfs`
 
-The kernel file under `/boot/` that is your running  kernel. e.g. `vmlinuz` or `bzImage`
+A list of sources (luet package, container image or local path) to install in the rootfs. The rootfs will be squashed to a `rootfs.squashfs` file
 
-### `initramfs.rootfs_file`
+### `iso.uefi`
 
-The initrd file under `/boot/` that has all the utils for the initramfs
+A list of sources (luet package, container image or local path) to install in the efi FAT image or partition.
 
-### `image_prefix`
+### `iso.image`
 
-ISO image prefix to use
+A list of sources (luet package, container image or local path) to install in ISO filesystem.
 
-### `image_date`
+### `iso.label`
 
-Boolean indicating if the output image name has to contain the date
+The label of the ISO filesystem. Defaults to `COS_LIVE`. Note this value is tied with the bootloader and kernel paramters to identify the root device.
 
-### `image_name`
+### `repositories`
 
-A string representing the ISO final image name
+A list of Luet package repositories
 
-### `arch`
```

Comment:
Note, there should also be an equivalent `repositories.arch`

Reply:
It is not there yet, this is something to revisit in the scope of build-disk I think. The repositories there are handled differently on per arch base and I think we should think about how to merge the disk manifest into a generic build manifest where repos stanza is shared for in top level structure, where Luet is defined. I guess one of the problems here is that the URI is different depending on the arch.

I think we need to agree on the expected behavior depending on the arch.

* Do we assume we can only build for current host archs? I mean no cross-arch builds.
* If cross archs builds are not supported we can think of making our defaults smarter and adapted according to the current host arch.
* If cross archs builds are not supported we can think of ignoring any repository from the repos list that does not match the architectoure (I am not sure if Luet already does it).
* So if the above is true I think we can make a `build-disk` manifest that only accepts repositories at top level instead of having a list per arch. So the repositories stanza could be shared between `build-iso` and `build-disk`. The default arch for a given repository if not specified should be host's arch.

IMHO I would definitely NOT support cross-arch builds for the moment. I'd do that via emulation instead, transparent from elemental.


What you think?

Reply:
I think that in the case of build-disk at least, its possible and easy to just cross build, after all its mainly copying the required packages into a dir and packing that into an image. So it would be good to have that in there, to speed up image building.

Not sure on iso build, but it also looks like the only diff between arches is the packages installed and the kernel naming (which is already covered in tour changes I think)? So again, it would greatly speed up the creation of different arches.

But yeah, having different repo names its a PITA.


I think luet does check the runtime.ARCH vs the list of repos in order to enable/disable them, but AFAIK that is outside of our implementation of luet, as I can easily use arm64 repos under x86_64. We could probably leverage that in our side to add all repos, but again, that would mean that we need qemu to build raw and iso, which is slowwwwwwww.

Reply:
See, this is a thing of beauty:

```
$ sudo bin/elemental build-disk -a arm64
INFO[2022-04-28T10:21:14+02:00] Starting elemental version v0.0.14           
INFO[2022-04-28T10:21:14+02:00] Repositories are empty, setting default value: quay.io/costoolkit/releases-green-arm64 
INFO[2022-04-28T10:21:14+02:00] Packages are empty, setting default values: [{Name:system/grub2-artifacts Target:root/grub2} {Name:recovery/cos-img Target:root/cOS} {Name:system/grub2-efi-image Target:efi} {Name:system/grub2-config Target:root}] 
INFO[2022-04-28T10:21:14+02:00] Building disk image type raw for arch arm64  
INFO[2022-04-28T10:23:22+02:00] Done! Image created at disk.raw              
```

it took 2 minutes and it built a nice, bootable arm64 image. I would hate to make this work on qemu :crying_cat_face:

Reply:
What about having an `--arch` flag in the build commands? So if no `--arch` is provided we use the hosts arch, if `--arch` is provided use whatever is given. Then we could make our default repos dependent on `arch` (given or from host) and filter installation repositories based on `arch`, so they all can be listed together. If a repository is given without an arch (via `--repo` flag or not including a the arch stanza in manifest) we just use whatever is defined by `--arch` flag or use host's arch.

So basically this means adding an Arch field in Config:

```golang
type Config struct {                                                            
    Logger          Logger                                                      
    Fs              FS                                                          
    Mounter         mount.Interface                                             
    Runner          Runner                                                      
    Syscall         SyscallInterface                                            
    CloudInitRunner CloudInitRunner                                             
    Luet            LuetInterface                                               
    Client          HTTPClient                                                  
    Cosign          bool         `yaml:"cosign,omitempty" mapstructure:"cosign"`
    CosignPubKey    string       `yaml:"cosign-key,omitempty" mapstructure:"cosign-key"`
    Repos           []Repository `yaml:"repositories,omit" mapstructure:"repositories"`
    // Architecture of the build artifacts
    Arch            string       `yaml:"arch,omit" mapstructure:"arch"`
} 
```
So we can set this and use that to filter repositories by arch. We should also include an Arch filed in `v1.Repository`.

I think this would solve this issue and allow a single arch independent repository list.